Python/ast.c

```diff
@@ -326,12 +326,234 @@ validate_expr(expr_ty exp, expr_context_ty ctx)
+                PyUnicode_CheckExact(literal)) {
+                return 1;
+            }
+            PyErr_SetString(PyExc_ValueError,
+                            "unexpected constant inside of a literal pattern");
+            return 0;
+        case Dict_kind:
+            keys = pattern->v.Dict.keys;
+            values = pattern->v.Dict.values;
+            length = asdl_seq_LEN(keys);
+            if (length != asdl_seq_LEN(values)) {
+                PyErr_SetString(
+                    PyExc_ValueError,
+                    "Dict doesn't have the same number of keys as values");
+                return 0;
+            }
+            expr_ty key, value;
+            for (i = 0; i < length; i++) {
+                key = asdl_seq_GET(keys, i);
+                value = asdl_seq_GET(values, i);
+                if (!key) {
+                    // The position of **rest is handled on the compiler
+                    if (value->kind == Name_kind) {
+                        if (!validate_expr(value, Store)) {
+                            return 0;
+                        }
+                    }
+                    else {
+                        PyErr_SetString(PyExc_ValueError,
+                                        "the double star pattern's target can "
+                                        "only be Name");
+                        return 0;
+                    }
+                }
+                else if (key->kind == Constant_kind ||
+                         key->kind == Attribute_kind ||
+                         key->kind == BinOp_kind) {
+                    if (!validate_pattern(key) || !validate_pattern(value)) {
+                        return 0;
+                    }
+                }
+                else {
+                    PyErr_SetString(
+                        PyExc_ValueError,
+                        "unexpected key type for a mapping pattern");
+                    return 0;
+                }
+            }
+            return 1;
+        case List_kind:
+        case Tuple_kind:
+            values = pattern->kind == List_kind ? pattern->v.List.elts
+                                                : pattern->v.Tuple.elts;
+            length = asdl_seq_LEN(values);
+
+            expr_ty element;
+            for (i = 0; i < length; i++) {
+                element = asdl_seq_GET(values, i);
+                if (element->kind == Starred_kind) {
+                    if (element->v.Starred.value->kind == Name_kind) {
+                        if (!validate_expr(element->v.Starred.value, Store)) {
+                            return 0;
+                        }
+                    }
+                    else {
+                        PyErr_SetString(PyExc_ValueError,
+                                        "the single star pattern's target can "
+                                        "only be Name");
+                        return 0;
+                    }
```

Comment:
This is a bit more readable for me:

```suggestion
                    if (element->v.Starred.value->kind != Name_kind) {
                        PyErr_SetString(PyExc_ValueError,
                                        "the single star pattern's target can "
                                        "only be Name");
                        return 0;
                    }
                    if (!validate_expr(element->v.Starred.value, Store)) {
                        return 0;
                    }
```